File: 03_Arrays/Easy/C++/07_union.cpp
// given -> 2 sorted array
// return -> union with distinct elements in sorted order


#include <bits/stdc++.h>
using namespace std;

vector < int > unionOfArray_brute(vector < int > a, vector < int > b) {
    set<int> st; // ordered set to preserve the order

    for (int i=0; i<a.size(); ++i){
        st.insert(a[i]);
    }

    for (int i=0; i<b.size(); ++i){
        st.insert(b[i]);
    }

    vector<int> myUnion;

    for (auto it: st){
        myUnion.push_back(it);
    }

    return myUnion;

}

vector < int > unionOfArray_Optimal(vector < int > a, vector < int > b) {
//    2-ptr Approach -> optimal
  int n1 = a.size();
  int n2 = b.size();

  vector<int> myUnion;

  int i = 0;
  int j = 0;

  while(i<n1 && j<n2){
      if(a[i] <= b[j]){
           if(myUnion.size() == 0 || myUnion.back() != a[i] ){
               myUnion.push_back(a[i]);
           }
           i++;
      }
      else {
           if(myUnion.size() == 0 || myUnion.back() != b[j] ){
               myUnion.push_back(b[j]);
           }
           j++;

      }
  }

   // if 2nd array is exhausted earlier
   // but 1st array still remains
  while(i<n1){
        if(myUnion.size() == 0 || myUnion.back() != a[i] ){
            myUnion.push_back(a[i]);
        }
        i++;
  }

   // if 1st array is exhausted earlier
   // but 2nd array still remains
  while(j<n2){
       if(myUnion.size() == 0 || myUnion.back() != b[j] ){
            myUnion.push_back(b[j]);
        }
        j++;
  }

   return myUnion;
}

int main(int argc, char* argv[]){
  vector<int> a;
  vector<int> b;

  int sizeA;
  cin >> sizeA;
  for (int i=0; i<sizeA; ++i) {
    int x;
    cin >> x;
    a.push_back(x); 
  }

  int sizeB;
  cin >> sizeB;
  for (int i=0; i<sizeB; ++i) {
    int x;
    cin >> x;
    b.push_back(x);
  }


  // vector<int> myUnion =  sortedArray_brute(a, b);
  vector<int> myUnion =  unionOfArray_Optimal(a, b);

  for (int i=0; i<myUnion.size(); ++i){
    cout << myUnion[i] << " ";
  }
return 0;
}

File: 03_Arrays/Easy/Python/07_union.py
# given -> 2 sorted arrays containing integers
# return -> union with distinct elements in sorted order

from typing import List

# class Solution(object): # -> Python 2 style
class Solution:
  # brute force -> using sets
  def union_of_Array_brute(self, arr_1: List[int], arr_2: List[int]) -> List[int]:
    # set insertion
    # set.add()
    st = set()

    for i in range(len(arr_1)):
      st.add(arr_1[i])
    
    for i in range(len(arr_2)):
      st.add(arr_2[i])

    ans = list(st)
    # ans = List(st)
    return ans

  # optimal -> 2-ptr approach
  def union_of_Array_optimal(self, arr_1: List[int], arr_2: List[int]) -> List[int]:
    n1 = len(arr_1)
    n2 = len(arr_2)

    ans = []

    i = 0
    j = 0

    while i < n1 and j < n2:
      if arr_1[i] <= arr_2[j]:
        if len(ans) == 0 or ans[-1] != arr_1[i]:
          ans.append(arr_1[i])
        i += 1
      
      else:
        if len(ans) == 0 or ans[-1] != arr_2[j]:
          ans.append(arr_2[j])
        j += 1

    
    while i < n1:
      if len(ans) == 0 or ans[-1] != arr_1[i]:
        ans.append(arr_1[i])
      i += 1 

    while j < n2:
      if len(ans) == 0 or ans[-1] != arr_2[j]:
        ans.append(arr_2[j])
      j += 1     
      

    return ans


size_1 = int(input("size 1: "))
size_2 = int(input("size 2: "))

arr_1 = list(map(int, input("Enter elements in array 1: ").strip().split()[ : size_1]))
arr_2 = list(map(int, input("Enter elements in array 2: ").strip().split()[ : size_2]))

my_union = Solution()

# print(my_union.union_of_Array_brute(arr_1, arr_2))
print(my_union.union_of_Array_optimal(arr_1, arr_2))